### PR TITLE
do opam file generation in the same order as `opam lint`

### LINKED ISSUE #2290: opam file generation: normalise resulting opam file
When generating opam files via dune-project, there's a big diff between the default order of opam fields generated by dune and those normalised by `opam lint <file> --normalise`

This is a minor hassle since the submitted version in dune-release could be run through `opam lint` before generating the release archive.  But if it's easy with the existing opam libraries vendored in dune, it would be nice to have the generated order match the opam linted field order.

### PR DESCRIPTION
**Pull request by avsm**
The logic for the field order used by `opam lint --normalise` is in opam-file and not opam-file-format. This is too much to vendor, so this commit hardcodes the field order based on the opam-file logic.

It won't change very often upstream so this shouldn't impose too much of a maintenance overhead.

Closes #2290

### CODE REVIEW

**Comment by rgrinberg** [comment]
Why do we do the sorting conditionally on the dune version?

My thinking is that if we've made an improvement to the opam generation, we should make it available for all versions.

**Comment by diml** [comment]
For files that are generated and committed in the source tree, it seems natural to tie the contents to the  dune language version.

File: src/opam_file.ml [hunk 1]
```diff
@@ -87,6 +87,51 @@ module Create = struct
...
+      "libraries";
+      "syntax";
+      "dev-repo";
+      "pin-depends";
+      "extra-files" ]
+    in
+    List.filter_map normal_field_order
+      ~f:(fun field ->
+        Option.map (List.assoc vars field)
+          ~f:(fun v -> field, v))
```
Comment (diml) [inline src/opam_file.ml]: In addition to sorting the fields, this is also silently filtering out fields dune doesn't know about. That seems a bit dodgy given that we allow the user to add arbitrary fields via the template file. I suggest to put all unknown fields at the end in their original order.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/opam_file.ml (hunk 1 above)
Comment (avsm) [reply src/opam_file.ml]: good catch! pushed a fix and some test cases for extension fields

**Comment by rgrinberg** [comment]
LGTM. I've tweaked the code a bit to do everything in a single pass of stable_sort. Probably an unnecessary micro optimization, but certainly not the first one in dune :)

**Comment by avsm** [comment]
Thanks! Cleanups to my hastily written code are always appreciated ;-)

**Pull request closed without merging**